runtime/lua/vim/filetype/detect.lua

```diff
@@ -1204,4 +1286,269 @@ end
+  ['^__libc_start_main'] = 'strace',
+  -- VSE JCL
+  ['^\\* $$ JOB\\>'] = { 'vsejcl', { vim_regex = true } },
+  ['^// *JOB\\>'] = { 'vsejcl', { vim_regex = true } },
+  -- TAK and SINDA
+  ['K & K  Associates'] = { 'takout', { start_lnum = 4 } },
+  ['TAK 2000'] = { 'takout', { start_lnum = 2 } },
+  ['S Y S T E M S   I M P R O V E D '] = { 'syndaout', { start_lnum = 3 } },
+  ['Run Date: '] = { 'takcmp', { start_lnum = 6 } },
+  ['Node    File  1'] = { 'sindacmp', { start_lnum = 9 } },
+  function(contents)
+    require('vim.filetype.detect').dns_zone(contents)
+  end,
+  -- Valgrind
+  ['^==%d+== valgrind'] = 'valgrind',
+  ['^==%d+== Using valgrind'] = { 'valgrind', { start_lnum = 3 } },
+  -- Go docs
+  ['PACKAGE DOCUMENTATION$'] = 'godoc',
+  -- Renderman Interface Bytestream
+  ['^##RenderMan'] = 'rib',
+  -- Scheme scripts
+  ['exec%s%+%S*scheme'] = { 'scheme', { start_lnum = 1, end_lnum = 2 } },
+  -- Git output
+  ['^\\(commit\\|tree\\|object\\) \\x\\{40,\\}\\>\\|^tag \\S\\+$'] = { 'git', { vim_regex = true } },
+  function(lines)
+    -- Gprof (gnu profiler)
+    if lines[1] == 'Flat profile:' and lines[2] == '' and lines[3]:find('^Each sample counts as .* seconds%.$') then
+      return 'gprof'
+    end
+  end,
+  -- Erlang terms
+  -- (See also: http://www.gnu.org/software/emacs/manual/html_node/emacs/Choosing-Modes.html#Choosing-Modes)
+  ['%-%*%-.*erlang.*%-%*%-'] = { 'erlang', { ignore_case = true } },
+  -- YAML
+  ['^%%YAML'] = 'yaml',
+  -- MikroTik RouterOS script
+  ['^#.*by RouterOS'] = 'routeros',
+  -- Sed scripts
+  -- #ncomment is allowed but most likely a false positive so require a space before any trailing comment text
+  ['^#n%s'] = 'sed',
+  ['^#n$'] = 'sed',
+}
+
+---@private
+-- File does not start with "#!".
+local function match_from_text(contents, path)
+  if contents[1]:find('^:$') then
+    -- Bourne-like shell scripts: sh ksh bash bash2
+    return M.sh(path, contents)
+  elseif matchregex('\n' .. table.concat(contents, '\n'), [[\n\s*emulate\s\+\%(-[LR]\s\+\)\=[ckz]\=sh\>]]) then
+    -- Z shell scripts
+    return 'zsh'
+  end
+
+  for k, v in pairs(patterns_text) do
+    if type(v) == 'string' then
+      -- Check the first line only
+      if contents[1]:find(k) then
+        return v
+      end
+    elseif type(v) == 'function' then
+      -- If filetype detection fails, continue with the next pattern
+      local ok, ft = pcall(v, contents)
+      if ok and ft then
+        return ft
+      end
+    else
+      local opts = type(v) == 'table' and v[2] or {}
+      if opts.start_lnum and opts.end_lnum then
+        assert(not opts.ignore_case, 'ignore_case=true is ignored when start_lnum is also present, needs refactor')
+        for i = opts.start_lnum, opts.end_lnum do
+          if not contents[i] then
+            break
+          elseif contents[i]:find(k) then
+            return v[1]
+          end
+        end
+      else
+        local line_nr = opts.start_lnum or 1
```

Comment:
```suggestion
        local line_nr = opts.start_lnum or 1
        if line_nr == -1 then line_nr = #contents end
```
(Probably can be collapsed into a more complicated assignment but grug brain does not like complexity.)

Reply:
Thanks, I'll push an update in a few minutes :)

Reply:
Too late ;) (I wanted to see what CI thinks of these suggestions.)

Reply:
(But feel free to drop my commit and replace with yours.)

Reply:
Made some tiny adjustments :) Filetype tests are passing locally.

Reply:
I'd remove the comment for the patterns (or move it to the top of the `gsub` chain) to stop Stylua from going haywire.